### PR TITLE
Ghastly desert + Spanish translations

### PR DESCRIPTION
**Pull request by emilyploszaj**
Additions
* Ghastly Desert - a biomes that generates like a vanilla desert but spookier
* Spooky Cactus - like a vanilla cactus but spookier, and restless
* Deader Bush - dead bush but more dead
* Spanish translations - like English but with more tildes

### CODE REVIEW

**Comment by Shnupbups** [comment]
Does Tainted Sandstone spawn in the Ghastly Desert?

**Comment by emilyploszaj** [comment]
> 
> 
> Does Tainted Sandstone spawn in the Ghastly Desert?

no, that was added after I worked on generation

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+			world.breakBlock(pos, true);
+		} else {
+			BlockPos pos2 = pos.up();
+			if (world.isAir(pos2)) {
+				int height;
+				for(height = 1; world.getBlockState(pos.down(height)).getBlock() == this; height++) {
+				}
+
+				if (height < 3) {
+					int age = (Integer)state.get(AGE);
```
Comment (Juuxel) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java]: Unnecessary cast here.

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+	public void onScheduledTick(BlockState state, World world, BlockPos pos, Random random) {
+		if (!state.canPlaceAt(world, pos)) {
+			world.breakBlock(pos, true);
+		} else {
+			BlockPos pos2 = pos.up();
+			if (world.isAir(pos2)) {
+				int height;
+				for(height = 1; world.getBlockState(pos.down(height)).getBlock() == this; height++) {
+				}
```
Comment (Juuxel) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java]: A while loop would be clearer than an empty for loop here.

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+			}
+		}
+	}
+
+	public VoxelShape getCollisionShape(BlockState state, BlockView view, BlockPos pos, EntityContext context) {
+		return COLLISION_SHAPE;
+	}
+
+	public VoxelShape getOutlineShape(BlockState state, BlockView view, BlockPos pos, EntityContext context) {
+	return OUTLINE_SHAPE;
```
Comment (Juuxel) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java]: Missing indentation here?

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+	public boolean canPlaceAt(BlockState state, ViewableWorld world, BlockPos pos) {
+		Iterator<Direction> iterator = Direction.Type.HORIZONTAL.iterator();
+		
+		Direction direction;
+		Material material;
+		do {
+			if (!iterator.hasNext()) {
+			Block block_1 = world.getBlockState(pos.down()).getBlock();
+			return (block_1 == SpookyBlocks.SPOOKY_CACTUS || block_1 == SpookyBlocks.TAINTED_SAND) && !world.getBlockState(pos.up()).getMaterial().isLiquid();
```
Comment (Juuxel) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java]: Missing indentation here as well.

**Review by Juuxel — CHANGES_REQUESTED** [review]
I left some comments on `SpookyCactusBlock`. Also, could you remove `_1` and `_2` from the variables that have them?

**Review by GiantNuker — APPROVED** [review]
I don't know spanish, but it looks like you followed the style conventions :)

File: src/main/java/com/fabriccommunity/spookytime/block/DeaderBushBlock.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import net.minecraft.world.BlockView;
+
+public class DeaderBushBlock extends PlantBlock {
+   public static final VoxelShape SHAPE = Block.createCuboidShape(2.0D, 0.0D, 2.0D, 14.0D, 13.0D, 14.0D);
+
+   public DeaderBushBlock(Block.Settings settings) {
+      super(settings);
+   }
+
+   public VoxelShape getOutlineShape(BlockState state, BlockView view, BlockPos pos, EntityContext context) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/DeaderBushBlock.java]: Can we add `@Override` annotations just for clarity? I'll make a quick review comment on every instance of this I see, just so we can keep track of them.

File: src/main/java/com/fabriccommunity/spookytime/block/DeaderBushBlock.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+      return SHAPE;
+   }
+
+   protected boolean canPlantOnTop(BlockState state, BlockView view, BlockPos pos) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/DeaderBushBlock.java]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+		COLLISION_SHAPE = Block.createCuboidShape(1.0D, 0.0D, 1.0D, 15.0D, 15.0D, 15.0D);
+		OUTLINE_SHAPE = Block.createCuboidShape(1.0D, 0.0D, 1.0D, 15.0D, 16.0D, 15.0D);
+	}
+	
+	public SpookyCactusBlock(Block.Settings settings) {
+		super(settings);
+		this.setDefaultState(this.stateFactory.getDefaultState().with(AGE, 0));
+	}
+
+	public void onScheduledTick(BlockState state, World world, BlockPos pos, Random random) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java:42]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+				SpookyCactusEntity entity = new SpookyCactusEntity(SpookyEntities.SPOOKY_CACTUS, world);
+				entity.setPosition(pos.getX() + 0.5f, pos.getY(), pos.getZ() + 0.5f);
+				entity.setCactusHeight(height);
+				entity.age = age;
+				world.spawnEntity(entity);
+			}
+		}
+	}
+
+	public VoxelShape getCollisionShape(BlockState state, BlockView view, BlockPos pos, EntityContext context) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java:84]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+		return COLLISION_SHAPE;
+	}
+
+	public VoxelShape getOutlineShape(BlockState state, BlockView view, BlockPos pos, EntityContext context) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java:89]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+		return OUTLINE_SHAPE;
+	}
+
+	public boolean isOpaque(BlockState state) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java:94]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+		return true;
+	}
+
+	public BlockState getStateForNeighborUpdate(BlockState state, Direction direction, BlockState state2, IWorld world, BlockPos pos, BlockPos pos2) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java:99]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+		if (!state.canPlaceAt(world, pos)) {
+			world.getBlockTickScheduler().schedule(pos, this, 1);
+		}
+
+		return super.getStateForNeighborUpdate(state, direction, state2, world, pos, pos2);
+	}
+
+	public boolean canPlaceAt(BlockState state, ViewableWorld world, BlockPos pos) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java:108]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+			direction = (Direction) iterator.next();
+			BlockState state2 = world.getBlockState(pos.offset(direction));
+			material = state2.getMaterial();
+		} while(!material.isSolid() && !world.getFluidState(pos.offset(direction)).matches(FluidTags.LAVA));
+
+		return false;
+	}
+
+	public void onEntityCollision(BlockState state, World world, BlockPos pos, Entity entity) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java:128]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+		entity.damage(DamageSource.CACTUS, 1.0F);
+	}
+
+	public BlockRenderLayer getRenderLayer() {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java:133]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+		return BlockRenderLayer.CUTOUT;
+	}
+
+	protected void appendProperties(StateFactory.Builder<Block, BlockState> builder) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java:138]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+		builder.add(AGE);
+	}
+
+	public boolean canPlaceAtSide(BlockState state, BlockView view, BlockPos pos, BlockPlacementEnvironment environment) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java:143]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+		CACTUS_HEIGHT = DataTracker.registerData(SpookyCactusEntity.class, TrackedDataHandlerRegistry.INTEGER);
+	}
+
+	public SpookyCactusEntity(EntityType<?> type, World world) {
+		super((EntityType<? extends MobEntityWithAi>) type, world);
+		setVelocity(0.0F, 0.5F, 0.0F);
+		calculateDimensions();
+	}
+
+	protected void initGoals() {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java:40]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+		this.goalSelector.add(0, new WanderAroundFarGoal(this, 1.0F));
+	}
+	
+	protected void initAttributes() {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java:45]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+		super.initAttributes();
+		this.getAttributeInstance(EntityAttributes.MAX_HEALTH).setBaseValue(10.0D);
+		this.getAttributeInstance(EntityAttributes.MOVEMENT_SPEED).setBaseValue(0.20000000298023224D);
+	 }
+
+	protected void initDataTracker() {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java:52]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/world/feature/DeaderBushFeature.java [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+import net.minecraft.world.gen.chunk.ChunkGeneratorConfig;
+import net.minecraft.world.gen.feature.DefaultFeatureConfig;
+import net.minecraft.world.gen.feature.Feature;
+
+public class DeaderBushFeature extends Feature<DefaultFeatureConfig> {
+   public DeaderBushFeature(Function<Dynamic<?>, ? extends DefaultFeatureConfig> function_1) {
+      super(function_1);
+   }
+
+	public boolean generate(IWorld world, ChunkGenerator<? extends ChunkGeneratorConfig> generator, Random random, BlockPos pos, DefaultFeatureConfig config) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/world/feature/DeaderBushFeature.java:21]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/world/feature/SpookyCactusFeature.java [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+import net.minecraft.world.gen.chunk.ChunkGeneratorConfig;
+import net.minecraft.world.gen.feature.DefaultFeatureConfig;
+import net.minecraft.world.gen.feature.Feature;
+
+public class SpookyCactusFeature extends Feature<DefaultFeatureConfig> {
+   public SpookyCactusFeature(Function<Dynamic<?>, ? extends DefaultFeatureConfig> function) {
+      super(function);
+   }
+
+	public boolean generate(IWorld world, ChunkGenerator<? extends ChunkGeneratorConfig> generator, Random random, BlockPos pos, DefaultFeatureConfig config) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/world/feature/SpookyCactusFeature.java:19]: `@Override`

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+public class SpookyCactusBlock extends Block {
+	public static final IntProperty AGE;
+	protected static final VoxelShape COLLISION_SHAPE;
+	protected static final VoxelShape OUTLINE_SHAPE;
+
+	static {
+		AGE = Properties.AGE_15;
+		COLLISION_SHAPE = Block.createCuboidShape(1.0D, 0.0D, 1.0D, 15.0D, 15.0D, 15.0D);
+		OUTLINE_SHAPE = Block.createCuboidShape(1.0D, 0.0D, 1.0D, 15.0D, 16.0D, 15.0D);
+	}
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java]: Might make sense to just same-line these without the static block.

File: src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+				}
+			}
+			if (world.getBlockState(pos.down(1)).getBlock() == SpookyBlocks.TAINTED_SAND && random.nextInt(2) == 0) {
+				int height = 0;
+				int age = 0;
+				for(height = 0; world.getBlockState(pos.up(height)).getBlock() == this; height++) {
+					age = world.getBlockState(pos.up(height)).get(AGE);
+					world.setBlockState(pos.up(height), Blocks.AIR.getDefaultState());
+				}
+				SpookyCactusEntity entity = new SpookyCactusEntity(SpookyEntities.SPOOKY_CACTUS, world);
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/block/SpookyCactusBlock.java:74]: I think we should have a newline before the entity just to split off the if-statement.

File: src/main/java/com/fabriccommunity/spookytime/client/render/SpookyCactusEntityRenderer.java [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+	@Override
+	public void render(SpookyCactusEntity entity, double x, double y, double z, float float_1, float float_2) {
+		GlStateManager.pushMatrix();
+		BlockRenderManager manager = MinecraftClient.getInstance().getBlockRenderManager();
+		BlockState state = SpookyBlocks.SPOOKY_CACTUS.getDefaultState();
+		this.bindTexture(SpriteAtlasTexture.BLOCK_ATLAS_TEX);
+		GlStateManager.translated(x, y, z);
+		GlStateManager.rotatef(entity.yaw, 0.0F, -1.0F, 0.0F);
+		GlStateManager.translated(-0.5F, 0.0F, 0.5F);
+		for(int i = 0; i < entity.getCactusHeight(); i++){
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/client/render/SpookyCactusEntityRenderer.java:35]: Could we newline out the if-statement on the front and back, just for clarity?

File: src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+
+@SuppressWarnings("unchecked")
+public class SpookyCactusEntity extends MobEntityWithAi {
+	public static final TrackedData<Integer> CACTUS_HEIGHT;
+	public BlockPos landPos;
+	public int age;
+	public int hop;
+
+	static {
+		CACTUS_HEIGHT = DataTracker.registerData(SpookyCactusEntity.class, TrackedDataHandlerRegistry.INTEGER);
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java]: I would just same-line this unless there's a particular reason for the static block.

File: src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+
+	@Override
+	public void tick() {
+		super.tick();
+		if (!world.isClient) {
+			List<Entity> collidingEntities = this.world.getEntities(Entity.class, this.getBoundingBox(), EntityPredicates.canBePushedBy(this));
+			for (Entity entity : collidingEntities) {
+				entity.damage(DamageSource.CACTUS, 1);
+			}
+			if (hop > 0) {
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java:66]: newline before hop statement

File: src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java [hunk 27]
```diff
@@ -0,0 +1,136 @@
...
+				double moveX = (x - landPos.getX() - 0.5F) / hop;
+				double moveZ = (z - landPos.getZ() - 0.5F) / hop;
+				float moveAngle = ((yaw - 90) % 360.0F) / hop;
+				setPositionAndAngles(x - moveX, y, z - moveZ, yaw - moveAngle, this.pitch);
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java:70]: same here, newline before hop check

File: src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+				if(world.getBlockState(getBlockPos().north()).getMaterial().isSolid()) return;
+				if(world.getBlockState(getBlockPos().south()).getMaterial().isSolid()) return;
+				if(world.getBlockState(getBlockPos().east()).getMaterial().isSolid()) return;
+				if(world.getBlockState(getBlockPos().west()).getMaterial().isSolid()) return;
+				for(int i = 0; i < getCactusHeight(); i++){
+					if(world.getBlockState(getBlockPos().up(i).north()).getMaterial().isSolid()) return;
+					if(world.getBlockState(getBlockPos().up(i).south()).getMaterial().isSolid()) return;
+					if(world.getBlockState(getBlockPos().up(i).east()).getMaterial().isSolid()) return;
+					if(world.getBlockState(getBlockPos().up(i).west()).getMaterial().isSolid()) return;
+				}
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java:91]: let's separate this statement out with newlines on front and end

File: src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+				this.goalSelector.disableControl(Control.TARGET);
+			}
+		}
+	}
+
+	@Override
+	public boolean damage(DamageSource source, float damage) {
+		if(source == DamageSource.CACTUS) return false;
+		return super.damage(source, damage);
+	}
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java:107]: Forgot a newline after the method body.

File: src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+		tag.putInt("CactusHeight", getCactusHeight());
+		tag.putInt("CactusAge", age);
+	}
+
+	@Override
+	public void readCustomDataFromTag(CompoundTag tag) {
+		super.readCustomDataFromTag(tag);
+		this.setCactusHeight(tag.getInt("CactusHeight"));
+		age = tag.getInt("CactusAge");
+	}
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java:121]: Here as well; just newline after the method body.

File: src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java [hunk 31]
```diff
@@ -0,0 +1,136 @@
...
+
+	@Override
+	public EntityDimensions getDimensions(EntityPose pose){
+		return EntityDimensions.changing(0.9F, 1.0F * getCactusHeight());
+	}
+
+	public float getPathfindingFavor(BlockPos pos, ViewableWorld world) {
+		return world.getBlockState(pos.down()).getBlock() == SpookyBlocks.TAINTED_SAND ? 10.0F : 1.0F;
+	}
+
```
Comment (Draylar) [inline src/main/java/com/fabriccommunity/spookytime/entity/SpookyCactusEntity.java]: Extra newline hiding in the back~

**Review by Draylar — APPROVED** [review]
I only have a bunch of style issues to address in my review. The content itself is great though, so I'm still approving.

:crab: style police is here :crab:

**Review by Shnupbups — APPROVED** [review]

**Pull request merged**